### PR TITLE
Incorporate ADAGE model parameter sweep

### PR DESCRIPTION
**Pull request by gwaygenomics**
Somewhat of a larger PR this time for one param sweep chunk.

Incorporates small changes to existing scripts to allow usage with ADAGE. Also adds specific ADAGE parameter sweep scripts and figures. The next PR will include the Jupyter notebook training the optimal model.

### CODE REVIEW

File: scripts/adage_pancancer.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+epochs = int(args.epochs)
+sparsity = float(args.sparsity)
+noise = float(args.noise)
+output_filename = args.output_filename
+
+original_dim = 5000
+latent_dim = 100
+
+# Random seed
+seed = int(np.random.randint(low=0, high=10000, size=1))
```
Comment (danich1) [inline scripts/adage_pancancer.py]: Why the completely random seed here?

File: scripts/vae_paramsweep.py [hunk 2]
```diff
@@ -29,60 +31,96 @@
 
 
 def get_param(param):
-    sweep = parameter_df.loc[param, 'sweep']
+    try:
```
Comment (danich1) [inline scripts/vae_paramsweep.py]: If you are going to do a try clause here, you might want to leave a message in the except block to let the user know an exception has occurred. Quietly handling this is fine, but will it break your code?

File: scripts/viz/param_sweep_adage_viz.R [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+# Sweep over measurements of kappa with batch_size and epochs as facets
+for (n in unique(param_melt_df$noise)) {
+  subset_param <- param_melt_df %>% dplyr::filter(noise == n)
+  
+  for (s in unique(param_melt_df$sparsity)) {
+    
+    subset_subset_param <- subset_param %>% dplyr::filter(sparsity == s)
+    # Remove label in `n` for plotting and filename
+    n <- gsub("noise: ", "", n)
+    s <- gsub("sparsity:", "", s)
```
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/viz/param_sweep_adage_viz.R]: :+1:

File: scripts/adage_pancancer.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+# Set hyper parameters
+learning_rate = float(args.learning_rate)
+batch_size = int(args.batch_size)
+epochs = int(args.epochs)
+sparsity = float(args.sparsity)
+noise = float(args.noise)
+output_filename = args.output_filename
+
+original_dim = 5000
```
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/adage_pancancer.py]: ah, it slipped in here again! I will fix in next commit

File: scripts/adage_pancancer.py (hunk 1 above)
Comment (gwaygenomics) [reply scripts/adage_pancancer.py]: I wanted to ensure that the splitting of training and testing data wasn't the same each time.

File: scripts/vae_paramsweep.py (hunk 2 above)
Comment (gwaygenomics) [reply scripts/vae_paramsweep.py]: good point. I will add in the next commit

File: scripts/adage_pancancer.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+test_set_percent = 0.1
+rnaseq_test_df = rnaseq_df.sample(frac=test_set_percent)
+rnaseq_train_df = rnaseq_df.drop(rnaseq_test_df.index)
+
+# Input place holder for RNAseq data with specific input size
+input_rnaseq = Input(shape=(original_dim, ))
+encoded_rnaseq = GaussianDropout(noise)(input_rnaseq)
+encoded_rnaseq_2 = Dense(latent_dim, activation='relu',
+                         activity_regularizer=l1(sparsity))(encoded_rnaseq)
+decoded_rnaseq = Dense(original_dim, activation='relu')(encoded_rnaseq_2)
```
[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/adage_pancancer.py]: Thanks for bringing this up - in initial development I tried several, all with varying results (Ideally I would have been documenting this, I know...) but after some quick prototyping I saw that `relu` had lower loss than both `sigmoid` and `tanh`.

However, I didn't look at downstream consequences of the model, including reconstruction. An issue I was having with `relu` is that some genes were simply not being reconstructed.

relu_gene_reconstruction

However, with `sigmoid` we no longer had this issue.

sigmoid_gene_reconstruction

Both models were trained with these params - which are different than my initial testing.

| param | optimal |
| :------: | :--------: |
| num_features | 5000 |
| encoding_dim | 100 |
| sparsity | 1e-6 |
| noise | 0.1 |
| epochs | 80 |
| batch_size | 50 |
| learning_rate | 0.9 |

[the remark this answers is not in the document]
Comment (gwaygenomics) [reply scripts/adage_pancancer.py]: Moral of the story is that I need to rerun the parameter sweep with a `sigmoid` loss.

Before I get this merged, I also need to investigate:

1. Is [`GaussianDropout` ](https://github.com/fchollet/keras/blob/master/keras/layers/noise.py#L52) doing what I think its doing
2. Is the `encoder` model and `decoder` model accurate? I may elect to simply use the weights learned by the autoencoder to obtain a latent representation instead of `encoder.predict()`. I think there may be something buggy going on here too.

**Comment by jaclyn-taroni** [comment]
@gwaygenomics is this ready for another look or are you still looking into `GaussianDropout`?

**Comment by gwaygenomics** [comment]
> is this ready for another look or are you still looking into GaussianDropout?

Still looking into it - I will write a message when the PR is set to review again

**Comment by gwaygenomics** [comment]
Going to track responses here (also for future reference).

In response to `GaussianDropout()` - I believe that its **not** doing what I think its doing, perhaps based on the name. The key component of `GaussianDropout()` is [here](https://github.com/fchollet/keras/blob/master/keras/layers/noise.py#L84-L86):

```python
return inputs * K.random_normal(shape=K.shape(inputs),
                                mean=1.0,
                                stddev=stddev)
```

Which is adding regularization, but not dropping out input weights. I should instead be using [`Dropout()`](https://github.com/fchollet/keras/blob/1d176e261469651067221448ee27c9b83abd1cc7/keras/layers/core.py#L72) after the input layer. I think this is more like the denoising autoencoder ADAGE is based from.

**Comment by gwaygenomics** [comment]
In response to the `encoder.predict()` comment:

I was previously building the encoder model with `encoder = Model(input_rnaseq, encoded_rnaseq_2)` and then building the latent features with `encoder.predict(rnaseq_df)`.  This was giving out a high abundance of zeroed out latent features. I thought this was an issue with the model, so I tested it by extracting the weights a different way with `weight_matrix = pd.DataFrame(autoencoder.get_weights()[0], index=rnaseq_df.columns)` and this did not produce zeroed latent features.

The key issue was the calling to `encoded_rnaseq_2 = Dense(latent_dim, activation='relu'...`, specifically, including the `relu` activation call in building this layer. `relu` is active when making the `encoder.predict()` call, which makes all negative features zero. So, to fix this, I will separate the activation call from the weight layer instance:

```python
# Build the Keras graph
input_rnaseq = Input(shape=(num_features, ))
encoded_rnaseq = Dropout(noise)(input_rnaseq)
encoded_rnaseq_2 = Dense(encoding_dim,
                         activity_regularizer=l1(sparsity))(encoded_rnaseq)
activation = Activation('relu')(encoded_rnaseq_2)
decoded_rnaseq = Dense(num_features, activation='sigmoid')(activation)

autoencoder = Model(input_rnaseq, decoded_rnaseq)
```

Separating this out solves the issue.

```python
g = (pd.DataFrame(np.array(results) - np.array(approx)).sum(axis=1)/ results.shape[0]).hist();
ax = g.get_figure()
ax.savefig('figures/testing_latent_feature.png')
```

testing_latent_feature

**Comment by gwaygenomics** [comment]
ok @jaclyn-taroni - ready for another look. Thanks!

**Comment by jaclyn-taroni** [comment]
@gwaygenomics can you comment on why the loss goes up in some of these figures? Is this a bug?

**Comment by gwaygenomics** [comment]
@jaclyn-taroni no bug, just `sparsity = 0.001` is too high. It zeros out too many weights for signal to be reconstructed and learning can never occur. I removed it from consideration in the overall final loss figure

**Comment by jaclyn-taroni** [comment]
@gwaygenomics - gotcha, thanks!

**Comment by gwaygenomics** [comment]
I should probably add the parameter sweep results for adage in this PR, since they're here for tybalt...

**Comment by gwaygenomics** [comment]
any other comments @jaclyn-taroni - if not, I will merge this in

**Comment by jaclyn-taroni** [comment]
@gwaygenomics  I will go through this shortly. Haven't taken a good look at all the changes, just wanted to check in about those plots first.

**Comment by gwaygenomics** [comment]
:+1:
